[Datahub] : Feature Catalog - User friendly labels

cmoinier opened the pull request:
### Description

This PR implements user friendly labels from feature catalog attributes, in the map feature details, the table columns and the charts dropdowns. It will attempt to match the feature's attribute "technical name" with the featureCatalog attribute name, and if a match exists, the featureCatalog "code" will be displayed instead of the feature's attribute "technical name". If there is no match (or no featureCatalog attributes), the technical name is kept.

### Architectural changes

none

### Screenshots

![image](https://github.com/user-attachments/assets/5d2f06ea-b09a-43f1-8e75-6a09c7dfa373)

![image](https://github.com/user-attachments/assets/76f46d1d-8f88-4ed6-8216-f2c5e3abf855)

### Quality Assurance Checklist

- [x] Commit history is devoid of any _merge commits_ and readable to facilitate reviews
- [x] If **new logic** ⚙️ is introduced: unit tests were added
- [ ] If **new user stories** 🤏 are introduced: E2E tests were added
- [ ] If **new UI components** 🕹️ are introduced: corresponding stories in Storybook were created
- [ ] If **breaking changes** 🪚 are introduced: add the `breaking change` label
- [ ] If **bugs** 🐞 are fixed: add the `backport <release branch>` label
- [ ] The [documentation website](docs) 📚 has received the love it deserves

<!--
Please only check items relevant to your contribution. Thank you very much for your time and efforts!
-->

### How to test

I have created a custom record to test it - it's hacky so I won't add it to the dump. Contact me to get the XML :)

AlitaBernachot commented on libs/ui/dataviz/src/lib/data-table/data-table.component.ts:
```diff
@@ -92,6 +93,7 @@ export class DataTableComponent implements OnInit, AfterViewInit, OnChanges {
 
   ngOnInit() {
     this.dataSource = new DataTableDataSource()
+    this.setProperties()
```
won't this `setProperties()` happen twice? first one is triggered by the setter.

cmoinier replied:
Indeed... The one in the setter is from a commit to fix the tests, so I guess it should be the right one. Will remove this one and try it out, thanks for pointing this out!

AlitaBernachot commented on libs/ui/dataviz/src/lib/data-table/data-table.component.ts:
```diff
@@ -61,14 +61,15 @@ export interface TableItemModel {
 })
 export class DataTableComponent implements OnInit, AfterViewInit, OnChanges {
+  _featureAttributes = []
+  @Input() set featureAttributes(value: []) {
```
could be nice to have a better typing

AlitaBernachot commented on libs/ui/dataviz/src/lib/data-table/data-table.component.ts:
```diff
@@ -115,6 +117,27 @@ export class DataTableComponent implements OnInit, AfterViewInit, OnChanges {
   }
 
+  setProperties() {
+    this.dataset_.properties.then((properties) => {
```
Not sure if the data-table component is really in charge of this label/code extracting process.

cmoinier replied:
I can definitely move it to the `table-view` component :)

AlitaBernachot commented on libs/feature/dataviz/src/lib/chart-view/chart-view.component.ts:
```diff
@@ -235,6 +238,28 @@ export class ChartViewComponent {
+          return p
+        })
+      })
+      .catch((error) => {
```
I don't know if this is due to these new changes but the chart is broken:

ok in int
![image](https://github.com/user-attachments/assets/b899dbd7-81b7-477c-ad10-6ae63f6c4a17)

ko in dev
![image](https://github.com/user-attachments/assets/6ea45542-a138-4b89-aae8-a696243e16fc)

AlitaBernachot replied:
Looks like I have the same result when switching to main.

cmoinier replied:
Oh right, I didn't even notice that this axis was down. Since this is also the case on main, it could be related to the recent layout changes (since we see it with the "old" layout on `int`. I will investigate because this should not be put on prod.

AlitaBernachot reviewed (approved):
Ok working fine for map.

For table i can see the Code is selected and the label is set as a fallback when no code exists. But the table content is empty (there is a warning). When i switch back to main branch, the table is displayed correctly. => i don't know if it's due to the added feature catalog?

For the chart, the Y axis is missing but also on main, this is weird, i guess this is not due to your code modifications.

AlitaBernachot commented on libs/ui/map/src/lib/components/feature-detail/feature-detail.component.ts:
```diff
@@ -14,11 +15,34 @@ const geometryKeys = ['geometry', 'the_geom']
+            key = matchingAttribute.code
+          }
+        }
+        acc[key] = this.feature.properties[p]
```
Looks like you have done the code extraction many times, can't this be refactor (if the modification is simple)?

cmoinier replied:
I would love to, but unfortunately the process is a bit different in all 3 cases, so I don't really see a way to harmonize it all...

cmoinier replied:
In the chart, we need the output to be a pair of value/label, in the data-table, a simple string, and here, we need to do key/value matching. I could put in common the matching part, but that would just be one line... I don't know if that's enough

cmoinier commented:
@AlitaBernachot With the latest commits, I made the following changes : 

- [ ] Fixed the behavior of the data-table --> I removed the part where the aggregation was set to `count` if there was no `newProp` because it was not waiting long enough for the `featureCatalog` attribute to load, and therefore hiding the Y axis as if there were no values in it (which is wrong). It felt okay to removed that because we are actually already checking for values in `yProperty` through `aggregationChoices` and `isCountAggregation`, so the check I removed was kind of creating a loop anyway.

AlitaBernachot commented:
With dataset: n_tri_lill_inondable_s_059 (updated with the xml you gave to me)
i have an error in console and the legend does not display. The legend is now displayed only when clicking on a feature.
![image](https://github.com/user-attachments/assets/e3cb7b6d-f94b-4ce3-badc-b41ea904a3f4)

AlitaBernachot commented on libs/feature/dataviz/src/lib/table-view/table-view.component.ts:
```diff
@@ -105,4 +103,23 @@ export class TableViewComponent {
+          }
+          return { value: p.name, label: p.name }
+        }
+        return { value: p.name, label: p.name }
```
Minor: l. 118 and l.120 are the same? can't this be refacto?

AlitaBernachot commented on libs/feature/dataviz/src/lib/chart-view/chart-view.component.ts:
```diff
@@ -78,6 +78,9 @@ export class ChartViewComponent {
   @Input() set link(value: DatasetOnlineResource) {
     this.currentLink$.next(value)
+    if (value) {
+      this.aggregation$.next('sum')
```
For my own understanding, why do we need to set aggr to 'sum' here?

cmoinier replied:
That was the whole issue with the graph, even below this PR's scope (which was emphazising it) : whenever aggregation is "count", then the Y dropdown isn't shown. And the aggregation is set to count when there are no suitable choices in the Y dropdown... As you can tell, this is a loop. So when the first source link was not producing an error and not populating the dropdown options OR actually not having Y choices (which is valid!) --> when this happened, the Y dropdown was hidden, which is fine. The issue is that, without initializing the aggregation when switching source, then the YChoices observable was not being triggered/subscribed to because the aggregation variable, which wasn't re-initialized, was hiding it (ngIf in the template) .

cmoinier commented on libs/feature/dataviz/src/lib/chart-view/chart-view.component.ts:
```diff
@@ -171,6 +174,9 @@ export class ChartViewComponent {
       if (!choices.find((choice) => choice.value === this.yProperty$.value)) {
         const newProp = choices[0]?.value || ''
+        if (!newProp && this.aggregation$.value !== 'count') {
+          this.aggregation$.next('count')
```
This is actually a revert of a previous commit, so this line is from the origin code :)
But I think this is because when there are no Y choices then you cannot really do any other aggregations than doing sums of the X axis data

AlitaBernachot commented:
I can see the table view but i cannot see the chart view: it looks like it does not refresh the chart with the WFS datasource ans stays on the csv source. When i test on main, the chart is displayed (with a broken displayed but it shows).
![image](https://github.com/user-attachments/assets/c4050823-6b62-4bca-b6bc-e1e5da74fe39)

cmoinier commented:
> With dataset: n_tri_lill_inondable_s_059 (updated with the xml you gave to me) i have an error in console and the legend does not display. The legend is now displayed only when clicking on a feature. ![image](https://private-user-images.githubusercontent.com/115457104/440382183-e3cb7b6d-f94b-4ce3-badc-b41ea904a3f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1MzI3NzAsIm5iZiI6MTc0NjUzMjQ3MCwicGF0aCI6Ii8xMTU0NTcxMDQvNDQwMzgyMTgzLWUzY2I3YjZkLWY5NGItNGNlMy1iYWRjLWI0MWVhOTA0YTNmNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxMTU0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOWViNjg5MTdiMjY5ZjU1NzEyZjRhOTgwYmY5ZDRjNjJmNDAyYjdkNGUzY2RiMzA0NWU2YThjYTJlOGYwMTY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0GkXaF3lnOxi5MJZzkke2LORzQOyX067z-XNp8RhAVw)
> 
> EDIT: I can reproduce on main, this is weird...

Yep, tested on the record I took this service from and I reproduce both this & the bug you reported on the graph, so I think this is the service's fault. I will create another record with a service & feature catalog that matches so you can properly test.

AlitaBernachot commented:
Tested with the new record you gave me => all good, thx for the work!

I don't know if it is also worth updating the label in the chart? Since it was not mentioned in the PR, I let you decide :)
![image](https://github.com/user-attachments/assets/28fc3013-bf88-4ace-8199-8bb54dc9bea4)
